.NET DI-ready localization content (#25495)
dotnet/docs: docs/core/extensions/localization.md

author: IEvangelist
ms.author: dapine
ms.date: 08/06/2021
---

# Localization in .NET

In this article, you will learn how to use the <xref:Microsoft.Extensions.Localization.IStringLocalizer%601> and <xref:Microsoft.Extensions.Localization.IStringLocalizerFactory> implementations. Localization is the process of translating an application's resources into localized versions for each culture that the application supports.

All of the example source code in this article relies on the [`Microsoft.Extensions.Localization`](https://www.nuget.org/packages/microsoft.extensions.localization) and [`Microsoft.Extensions.Hosting`](https://www.nuget.org/packages/microsoft.extensions.hosting) NuGet packages. For more information on hosting, see [.NET Generic Host](generic-host.md).

## Resource files

The primary mechanism for isolating localizable strings is with **resource files**. A resource file is an XML file with the *.resx* file extension. Resource files are translated prior to the execution of the consuming application &mdash; in other words, they represent translated content at rest. A resource file name most commonly contains a locale identifier, and takes on the following form:

> **`{FullTypeName}[.Locale].resx`**

Reviewer: The CG says to use angle brackets for placeholders, and perhaps italics too. https://review.docs.microsoft.com/en-us/help/contribute/code-in-docs?branch=master#placeholders

Author: Ah, good to know. Fixed

Where:

- The `{FullTypeName}` represents localizable resources for a specific type.
- The optional `[.Locale]` represents the locale of the of the resource file contents.

### Specifying locales

> The locale at a bare minimum should define the language, but it can also define the culture (dialect), and even the country. These segments are commonly delimited by the `-` character. With the added specificity of a culture, the "culture fallback" rules are applied where best matches are prioritized. The locale should map to a well-known language tag, see <xref:System.Globalization.CultureInfo.Name?displayProperty=nameWithType>.

Reviewer: ```suggestion
The locale should define the language, at a bare minimum, but it can also define the culture (dialect), and even the country. These segments are commonly delimited by the `-` character. With the added specificity of a culture, the "culture fallback" rules are applied where best matches are prioritized. The locale should map to a well-known language tag. For more information, see <xref:System.Globalization.CultureInfo.Name?displayProperty=nameWithType>.
```

| *MessageService.resx*            |                               |              |

_<sup>†</sup>  The default dialect for the language._

When your app is running with the <xref:System.Globalization.CultureInfo.CurrentCulture?displayProperty=nameWithType> set to a culture of `"sr-Cyrl-RS"` localization attempts to resolve files in the following order:

1. *MessageService.sr-Cyrl-RS.resx*
1. *MessageService.sr-Cyrl.resx*
1. *MessageService.sr.resx*
1. *MessageService.resx*

However, if your app was running with the <xref:System.Globalization.CultureInfo.CurrentCulture?displayProperty=nameWithType> set to a culture of `"sr-Latn-BA"` localization attempts to resolve files in the following order:

1. *MessageService.sr-Latn-BA.resx*
1. *MessageService.sr-Latn.resx*
1. *MessageService.sr.resx*
1. *MessageService.resx*

The "culture fallback" rule will ignore locales when there are no corresponding matches, meaning resource file number four is selected if it's unable to find a match. If the culture was set to `"fr-FR"`, localization would end up falling to the *MessageService.resx* file which can be problematic.

### Resource lookup

> Resource files are automatically resolved as part of a lookup routine. If your project file name is different than the root namespace of your project, the assembly name might differ. This can prevent resource lookup from being otherwise successful. To address this mismatch, use the <xref:Microsoft.Extensions.Localization.RootNamespaceAttribute> to provide a hint to the localization services. When provided it is used during resource lookup.

Reviewer: ```suggestion
Resource files are automatically resolved as part of a lookup routine. If your project file name is different than the root namespace of your project, the assembly name might differ. This can prevent resource lookup from being otherwise successful. To address this mismatch, use the <xref:Microsoft.Extensions.Localization.RootNamespaceAttribute> to provide a hint to the localization services. When provided, it is used during resource lookup.
```

> The example project is named *example.csproj*, which creates a *example.dll* and *example.exe* &mdash; however; the `Localization.Example` namespace is used. Apply an `assembly` level attribute to correct this mismatch:

Reviewer: ```suggestion
The example project is named *example.csproj*, which creates a *example.dll* and *example.exe* &mdash; however, the `Localization.Example` namespace is used. Apply an `assembly` level attribute to correct this mismatch:
```

:::code source="snippets/localization/example/Program.cs" range="11":::

## Register localization services

> To register localization services call one of the <xref:Microsoft.Extensions.DependencyInjection.LocalizationServiceCollectionExtensions.AddLocalization%2A> extension methods durning the configuration of services. This will enable dependency injection (DI), of the following types:

Reviewer: ```suggestion
To register localization services, call one of the <xref:Microsoft.Extensions.DependencyInjection.LocalizationServiceCollectionExtensions.AddLocalization%2A> extension methods during the configuration of services. This will enable dependency injection (DI) of the following types:
```

- <xref:Microsoft.Extensions.Localization.IStringLocalizer%601?displayProperty=fullName>
- <xref:Microsoft.Extensions.Localization.IStringLocalizerFactory?displayProperty=fullName>

### Configure localization options

> The <xref:Microsoft.Extensions.DependencyInjection.LocalizationServiceCollectionExtensions.AddLocalization(Microsoft.Extensions.DependencyInjection.IServiceCollection,System.Action{Microsoft.Extensions.Localization.LocalizationOptions})> overload, accepts a `setupAction` parameter of `Action<LocalizationOptions>` type. This allows for the configuring of localization options.

Reviewer: ```suggestion
The <xref:Microsoft.Extensions.DependencyInjection.LocalizationServiceCollectionExtensions.AddLocalization(Microsoft.Extensions.DependencyInjection.IServiceCollection,System.Action{Microsoft.Extensions.Localization.LocalizationOptions})> overload accepts a `setupAction` parameter of type `Action<LocalizationOptions>`. This allows you to configure localization options.
```

- <xref:Microsoft.Extensions.Localization.IStringLocalizerFactory>

To create a message service that is capable of returning localized strings, consider the following `MessageService`:

:::code source="snippets/localization/example/MessageService.cs" highlight="8,10-11,16":::

In the preceding C# code:

- A `IStringLocalizer<MessageService> _localizer` field is declared.
- The constructor takes a `IStringLocalizer<MessageService>` parameter and assigns it to the `_localizer` field.
- The `GetGreetingMessage` method invokes the <xref:Microsoft.Extensions.Localization.IStringLocalizer.Item(System.String)?displayProperty=nameWithType> passing `"GreetingMessage"` as an argument.

The `IStringLocalizer` also supports parameterized string resources, consider the following `ParameterizedMessageService`:

:::code source="snippets/localization/example/ParameterizedMessageService.cs" highlight="9,11-12,17":::

In the preceding C# code:

- A `IStringLocalizer _localizer` field is declared.
- The constructor takes a `IStringLocalizerFactory` parameter, which is used to create a `IStringLocalizer` from the `ParameterizedMessageService` type, and assigns it to the `_localizer` field.

> - The `GetFormattedMessage` method invokes the <xref:Microsoft.Extensions.Localization.IStringLocalizer.Item(System.String,System.Object[])?displayProperty=nameWithType> passing `"DinnerPriceFormat"`, a `dateTime`, and the `dinnerPrice` as arguments.

Reviewer: ```suggestion
- The `GetFormattedMessage` method invokes <xref:Microsoft.Extensions.Localization.IStringLocalizer.Item(System.String,System.Object[])?displayProperty=nameWithType>, passing `"DinnerPriceFormat"`, a `dateTime` object, and `dinnerPrice` as arguments.
```

> [!IMPORTANT]
> The `IStringLocalizerFactory` isn't required. Instead, it is preferred for consuming services to require the <xref:Microsoft.Extensions.Localization.IStringLocalizer%601>.

Both <xref:Microsoft.Extensions.Localization.IStringLocalizer.Item%2A?displayProperty=nameWithType> indexers return a <xref:Microsoft.Extensions.Localization.LocalizedString>, which have [implicit conversions](xref:Microsoft.Extensions.Localization.LocalizedString.op_Implicit(Microsoft.Extensions.Localization.LocalizedString)~System.String) to `string?`.

## Put it all together

> To exemplify an app using both message services, with localization and resource files consider the following *Program.cs* file:

Reviewer: ```suggestion
To exemplify an app using both message services, along with localization and resource files, consider the following *Program.cs* file:
```

:::code source="snippets/localization/example/Program.cs":::

In the preceding C# code:

- The <xref:Microsoft.Extensions.Localization.RootNamespaceAttribute> sets `"Localization.Example"` as the root namespace.
- The <xref:System.Console.OutputEncoding?displayProperty=nameWithType> is assigned to <xref:System.Text.Encoding.Unicode?displayProperty=nameWithType>.
- When a single argument is passed to `args`, the <xref:System.Globalization.CultureInfo.CurrentCulture?displayProperty=nameWithType> and <xref:System.Globalization.CultureInfo.CurrentUICulture?displayProperty=nameWithType> are assigned the result of <xref:System.Globalization.CultureInfo.GetCultureInfo(System.String)?displayProperty=nameWithType> given the `arg[0]`.

> - The host is created with [defaults](generic-host.md#default-builder-settings), localization services, `MessageService`, and `ParameterizedMessageService` are all added to the `IServiceCollection` for DI.

Reviewer: This sentence is grammatically incorrect. I'm not sure the intended meaning so not supplying a suggestion here.

> - Logging is configured to ignore any log level lower than a warning to remove noise.

Reviewer: ```suggestion
- To remove noise, logging is configured to ignore any log level lower than a warning.
```

- The `MessageService` is resolved from the `IServiceProvider` instance and its resulting message is logged.
- The `ParameterizedMessageService` is resolved from the `IServiceProvider` instance and its resulting formatted message is logged.

> Each of the `*MessageService` classes, define a set of *.resx* files, each with a single entry. Here is the example content for the `MessageService` resource files, starting with *MessageService.resx*:

Reviewer: ```suggestion
Each of the `*MessageService` classes defines a set of *.resx* files, each with a single entry. Here is the example content for the `MessageService` resource files, starting with *MessageService.resx*:
```

:::code language="xml" source="snippets/localization/example/MessageService.resx" range="1-2,120-123":::

*MessageService.sr-Cyrl-RS.resx*:

:::code language="xml" source="snippets/localization/example/MessageService.sr-Cyrl-RS.resx" range="1-2,120-123":::

*MessageService.sr-Latn.resx*:

:::code language="xml" source="snippets/localization/example/MessageService.sr-Latn.resx" range="1-2,120-123":::

Here is the example content for the `ParameterizedMessageService` resource files, starting with *ParameterizedMessageService.resx*:

:::code language="xml" source="snippets/localization/example/ParameterizedMessageService.resx" range="1-2,120-123":::

*ParameterizedMessageService.sr-Cyrl-RS.resx*:

:::code language="xml" source="snippets/localization/example/ParameterizedMessageService.sr-Cyrl-RS.resx" range="1-2,120-123":::

*ParameterizedMessageService.sr-Latn.resx*:

:::code language="xml" source="snippets/localization/example/ParameterizedMessageService.sr-Latn.resx" range="1-2,120-123":::

> [!TIP]

> > All of the resource file XML comments, schema and `<resheader>` elements are intentionally omitted for brevity.

Reviewer: ```suggestion
> All of the resource file XML comments, schema, and `<resheader>` elements are intentionally omitted for brevity.
```

### Example runs

> The following example runs show the various localized outputs, given targeted locales &mdash; consider `"sr-Latn"`:

Reviewer: ```suggestion
The following example runs show the various localized outputs, given targeted locales.

Consider `"sr-Latn"`:
```

```dotnetcli
dotnet run --project .\example\example.csproj sr-Latn

warn: Localization.Example[0]
      Zdravo prijatelji, ".NET" developer zajednica je uzbuđena što vas vidi ovde!
warn: Localization.Example[0]
      U utorak, 03. avgust 2021. moja večera je koštala 37,63 ¤.
```

When omitting an argument to the [.NET CLI to run](../tools/dotnet-run.md) the project, the default system culture is used &mdash; in this case `"en-US"`:

```dotnetcli
dotnet run --project .\example\example.csproj

warn: Localization.Example[0]
      Hi friends, the ".NET" developer community is excited to see you here!
warn: Localization.Example[0]
      On Tuesday, August 3, 2021 my dinner cost $37.63.
```

When passing `"sr-Cryl-RS"`, the correct corresponding resource files are found and the localization applied:

```dotnetcli
dotnet run --project .\example\example.csproj sr-Cryl-RS

warn: Localization.Example[0]
      Здраво пријатељи, ".NЕТ" девелопер заједница је узбуђена што вас види овде!
warn: Localization.Example[0]
      У уторак, 03. август 2021. моја вечера је коштала 38 RSD.
```

> The sample application does not provide resource files for `"fr-CA"`, but when called with that culture, the non-locale resource files are fallen back to.

Reviewer: ```suggestion
The sample application does not provide resource files for `"fr-CA"`, but when called with that culture, the non-localized resource files are used.
```

---
title: Localization in .NET

> description: Learn how to use the IStringLocalizer and IStringLocalizerFactory implementations with dependency injection in your .NET workloads.

Reviewer: Seems like this should be expanded a bit to depict more general localization info.

author: IEvangelist
ms.author: dapine
ms.date: 08/10/2021
helpviewer_keywords:
  - "culture, localization"
  - "application development [.NET], localization"
  - "globalization [.NET], localization"
  - "code blocks"
  - "international applications [.NET], localization"
  - "global applications, localization"
  - "world-ready applications, localization"
  - "user interface blocks"
  - "localization [.NET], about localization"
  - "localizing resources"
ms.assetid: 49d520d7-92d7-44ee-bb24-8b615db1d41b
---

# Localization in .NET

> In this article, you will learn how to use the <xref:Microsoft.Extensions.Localization.IStringLocalizer%601> and <xref:Microsoft.Extensions.Localization.IStringLocalizerFactory> implementations. Localization is the process of translating an application's resources into localized versions for each culture that the application will support. You should proceed to the localization step only after completing the [Localizability review](localizability-review.md) step to verify that the globalized application is ready for localization.

Reviewer: ```suggestion
Localization is the process of translating an application's resources into localized versions for each culture that the application will support. You should proceed to the localization step only after completing the [Localizability review](localizability-review.md) step to verify that the globalized application is ready for localization.
```

Reviewer: Didn't seem to fit as the first sentence, so i moved it down a few paragraphs.

An application that is ready for localization is separated into two conceptual blocks: a block that contains all user interface elements and a block that contains executable code. The user interface block contains only localizable user-interface elements such as strings, error messages, dialog boxes, menus, embedded object resources, and so on for the neutral culture. The code block contains only the application code to be used by all supported cultures. The common language runtime supports a satellite assembly resource model that separates an application's executable code from its resources. For more information about implementing this model, see [Resources in .NET](resources.md).

For each localized version of your application, add a new satellite assembly that contains the localized user interface block translated into the appropriate language for the target culture. The code block for all cultures should remain the same. The combination of a localized version of the user interface block with the code block produces a localized version of your application.

> All of the example source code in this article relies on the [`Microsoft.Extensions.Localization`](https://www.nuget.org/packages/microsoft.extensions.localization) and [`Microsoft.Extensions.Hosting`](https://www.nuget.org/packages/microsoft.extensions.hosting) NuGet packages. For more information on hosting, see [.NET Generic Host](generic-host.md).

Reviewer: ```suggestion
In this article, you will learn how to use the <xref:Microsoft.Extensions.Localization.IStringLocalizer%601> and <xref:Microsoft.Extensions.Localization.IStringLocalizerFactory> implementations. All of the example source code in this article relies on the [`Microsoft.Extensions.Localization`](https://www.nuget.org/packages/microsoft.extensions.localization) and [`Microsoft.Extensions.Hosting`](https://www.nuget.org/packages/microsoft.extensions.hosting) NuGet packages. For more information on hosting, see [.NET Generic Host](generic-host.md).
```
